### PR TITLE
DEV-1538: probe actual SQLite stored values to widen INT to DOUBLE/TEXT

### PR DESCRIPTION
**Pull request by ZmeiGorynych**
## Summary

- SQLite's declared column types are affinity hints, not constraints — a column declared `INTEGER` can store mixed `INTEGER` / `REAL` / `TEXT` / `BLOB` per row. Trusting the declared type at ingest produces silent `CAST(... AS INTEGER)` truncation downstream (e.g. `AVG(0.99, 0.943, …)` returning ~0 instead of ~0.9 on the LiveSQLBench `vaccine` benchmark).
- New `slayer/sql/sqlite_introspect.py::probe_sqlite_integer_column` runs a per-column value-level probe (`typeof()`, `ROUND(col) <> col`, BLOB count, plus a Python `float()` coerce of distinct text values) over a `LIMIT 100_001` sample and returns the appropriate `DataType` — `INT` only when there's positive integer evidence (and the sample isn't saturated), `DOUBLE` for any REAL / non-integral / coercible-text storage, `TEXT` for BLOB or non-coercible / non-finite text or saturated distinct-text. Empty / all-NULL / saturated samples return `None` (no evidence) so callers keep the SA-derived type.
- Wired into four places (gated on `sa_engine.dialect.name == "sqlite"`):
  1. **Fresh ingest** (`ingest_datasource`, `introspect_table_to_model`) — widens persisted column types at first look.
  2. **Idempotent re-ingest** (`_additive_merge_existing`) — widens persisted `INT` → `DOUBLE` / `TEXT` based on the fresh probe; only the auto-default `NumberFormat(INTEGER)` is flipped to `FLOAT` / cleared; custom formats preserved with INFO log. `ModelAddition.widened_columns: List[str]` surfaces the events through the CLI renderer.
  3. **Load-time refinement** (`refine_dict_with_live_schema`) — SQLite is now probe-gated in **both** directions: DEV-1538 widening (INT → DOUBLE / TEXT) AND DEV-1361 narrowing (DOUBLE → INT, only when the probe positively certifies INT). DEV-1361 declared-type-driven narrowing is removed on SQLite.
  4. **`validate_models`** — runs the probe per persisted `INT` base column on SQLite and merges drift entries into `sql_table_diffs` **before** the cascade fixed-point walk so dependent measures, derived columns, joins, and filters cascade-drop in the same pass. Reason text: `"SQLite affinity probe widened <model>.<col> from INT to <verdict>; re-run slayer ingest to recreate with the correct type."`

Pre-DEV-1538 v7 SQLite models with the wrong `INT` type are **not** auto-repaired on load (full-table-scan per column on every `storage.get_model` would be prohibitive). Re-ingest is the auto-heal: `slayer ingest` or `slayer serve --ingest-on-startup`.

Non-SQLite engines (Postgres, MySQL, DuckDB, ClickHouse, SQL Server) skip the probe entirely.

## Test plan

- [x] 61 new tests across `tests/test_sqlite_probe.py` (helper), `tests/test_ingestion.py` (fresh ingest), `tests/test_idempotent_ingestion.py` (re-ingest widening), `tests/test_storage_type_refinement.py` (load-time probe-gated refinement), `tests/test_validate_models.py` (drift detection + cascade), plus one end-to-end integration test in `tests/integration/test_integration.py` that ingests a vaccine-style fixture and confirms `AVG(tempstabidx)` returns ~1.1 instead of zero-truncated.
- [x] Full non-integration suite green (3826 passed, 43 skipped) after merge from `main`.
- [x] Linter clean (`poetry run ruff check slayer/ tests/`).
- [ ] Manual smoke test on the LiveSQLBench `vaccine` DB (re-run the failing trajectory) — recommended before merging.

## Codex review

Two rounds of Codex review folded in:
- Round 1 (plan): decision order put TEXT check before REAL (so REAL + non-coercible TEXT → TEXT), BLOB always → TEXT, saturated samples return None (conservative), validate_models drift merged into cascade state pre-walk, NaN/Inf guard via `math.isfinite()`, format-flip only on auto-default `NumberFormat(INTEGER)`.
- Round 2 (tests): added constants assertions for `PROBE_SCAN_CAP` / `COERCE_DISTINCT_LIMIT`, BLOB tests, distinct-text saturation vs. low-distinct-text-with-many-rows split, coerce-query-failure-distinct-from-main-failure test, schema-qualified ATTACH DATABASE test, `validate_models` TEXT-drift case, cascade-to-derived-column and cascade-to-dependent-measure tests, tightened reason-text assertion, dotted-alias skip test.

## Linear

[DEV-1538](https://linear.app/motley-ai/issue/DEV-1538/sqlite-ingestion-probe-actual-stored-values-before-trusting-declared)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * SQLite integer-declared columns are now value-probed at ingest/load and can be auto-widened to decimal or text; widen events are surfaced in ingest/idempotent-reingest results, validation output, and CLI summaries. CLI/type-migration skips missing-datasource cases for SQLite widening with a warning while still erroring for other refinements.

* **Documentation**
  * Added guidance on SQLite affinity probing, probe sampling/decision rules, re-ingest behavior, and schema-drift handling.

* **Tests**
  * Extensive new tests covering probing logic, widening scenarios, idempotent re-ingest, validation cascades, CLI output, and edge cases.

### CODE REVIEW

File: docs/concepts/ingestion.md [hunk 1]
```diff
@@ -51,6 +51,23 @@ FK columns from referenced tables are excluded from the source model to avoid re
 
 All models use `sql_table` (the source table) plus `joins` (direct FK joins only, storing source/target column pairs). Multi-hop JOINs are resolved dynamically at query time by walking the join graph.
 
+### SQLite affinity probing
+
+SQLite's declared column types are affinity hints, not strict constraints: a column declared `INTEGER` can store `INTEGER`, `REAL`, `TEXT`, or `BLOB` values per row. To prevent silent truncation downstream (a column declared `INTEGER` but actually storing `0.99` would cast to `0` and break `AVG`/`SUM` results), SLayer runs an additional value-level probe on SQLite ingestion for every column the inspector reports as `INTEGER`-affinity.
+
+The probe samples up to **100,000 rows** (configurable via `slayer.sql.sqlite_introspect.PROBE_SCAN_CAP`) and decides per column:
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply docs/concepts/ingestion.md]: Addressed in commit follow-up: updated `docs/concepts/ingestion.md` to say "up to `PROBE_SCAN_CAP + 1` rows (100,001 by default)" and explained the `+1` as the saturation-detection guard.

File: docs/concepts/schema-drift.md [hunk 2]
```diff
@@ -130,6 +130,26 @@ models and the result is merged into `IdempotentIngestResult.to_delete`.
...
 and the validator — excluded tables are not touched in either direction.
 
+## SQLite affinity-probe drift
+
+On SQLite, `validate_models` additionally runs the value-level affinity
+probe (see [Ingestion → SQLite affinity probing](ingestion.md#sqlite-affinity-probing))
+against every persisted base column with `type = INT`. When the probe
+disagrees with the persisted type — the live storage class is `REAL`,
+non-coercible `TEXT`, `BLOB`, or `INTEGER` with non-integral values —
+`validate_models` emits an `EditModelDelete` removing the column with
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply docs/concepts/schema-drift.md]: Addressed: reworded `docs/concepts/schema-drift.md` to say "non-integral numeric evidence" instead of the impossible "INTEGER with non-integral values" phrasing. The `ROUND(col) <> col` term in the probe is defence-in-depth — in practice SQLite stores non-integral values as REAL anyway, but the wording is now accurate.

File: slayer/engine/schema_drift.py [hunk 3]
```diff
@@ -1756,14 +1759,168 @@ def _resolve_live_table(
...
+    from slayer.sql.sqlite_introspect import probe_sqlite_integer_column
+
+    if model.sql_table is None:
+        return []
+
+    if "." in model.sql_table:
+        schema_name, _, table_name = model.sql_table.partition(".")
+        schema_name = schema_name or None
+    else:
+        schema_name, table_name = None, model.sql_table
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/schema_drift.py]: Addressed: `_sqlite_probe_int_drift_for_model` now accepts `default_schema` and uses it for unqualified `model.sql_table` names. `_sqlite_probe_drifts_for_models` passes `datasource.schema_name` through. Same `default_schema` plumbing applied in `slayer/storage/type_refinement.py::_parse_sql_table_with_default_schema` for the load-time refinement path.

File: slayer/sql/sqlite_introspect.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+          (conservatively widened — we can't prove all distinct values
+          coerce).
+       c. Any text value's ``float()`` raises or result is not finite →
+          :class:`DataType.TEXT`.
+       d. Every distinct text value coerces to a finite float →
+          :class:`DataType.DOUBLE`.
+    4. ``n_real > 0`` OR ``n_non_integral > 0`` → :class:`DataType.DOUBLE`.
+    5. All-zero counts:
+       a. Row sample saturated → log WARNING, return ``None``.
+       b. Otherwise (empty / all-NULL / all-INTEGER) → :class:`DataType.INT`.
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/sql/sqlite_introspect.py]: Addressed: the decision-tree docstring on `probe_sqlite_integer_column` (and the section 5 inline comment) now match the implementation — empty / all-NULL samples return `None` ("no non-NULL evidence"), and the INT verdict only fires when there's positive integer evidence.

File: slayer/sql/sqlite_introspect.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        return DataType.TEXT
+
+    # 2. TEXT presence forces the coerce branch BEFORE the REAL branch so
+    #    a column with REAL + non-coercible TEXT widens to TEXT, not DOUBLE.
+    if n_text > 0:
+        distinct_limit_plus_one = COERCE_DISTINCT_LIMIT + 1
+        coerce_sql = sa.text(
+            f"SELECT DISTINCT {qcol} FROM {qtable} "
+            f"WHERE typeof({qcol}) = 'text' "
+            f"LIMIT {distinct_limit_plus_one}"
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/sql/sqlite_introspect.py]: Addressed: the distinct-text coerce query now wraps the source in the same `LIMIT PROBE_SCAN_CAP + 1` row-sample subquery as the main probe. The whole probe stays bounded.

File: slayer/storage/type_refinement.py [hunk 6]
```diff
@@ -1,25 +1,41 @@
...
+DEV-1538 splits this module's behavior by datasource type:
+
+* **SQLite**: skip the declared-type-driven narrowing entirely. Instead,
+  for every persisted INT base column, run
+  :func:`slayer.sql.sqlite_introspect.probe_sqlite_integer_column` against
+  the live storage classes and widen the persisted type to DOUBLE / TEXT
+  when the probe disagrees. The auto-default integer ``format`` is also
+  flipped (FLOAT for DOUBLE, cleared for TEXT); user-set custom formats
+  are preserved verbatim and an INFO log line is emitted as a hint.
+* **Non-SQLite**: existing DEV-1361 narrowing unchanged.
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/storage/type_refinement.py:19]: Addressed: updated the module docstring and `refine_dict_with_live_schema`'s SQLite/Non-SQLite sections to describe the probe-verified bidirectional behavior (widen INT and probe-verified narrow DOUBLE).

File: tests/test_sqlite_probe.py [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+
+class TestProbeTextCoerce:
+    def test_text_coerce_widens_to_double(self, conn) -> None:
+        from slayer.sql.sqlite_introspect import probe_sqlite_integer_column
+
+        conn.execute(sa.text('CREATE TABLE t (v INTEGER)'))
+        # All TEXT, all numeric-coercible to finite floats.
+        _insert_typed(conn, "t", "v", ["1", "2.5", "1e3", "-7.2", "  4  "])
+        verdict = probe_sqlite_integer_column(conn=conn, table="t", column="v")
+        assert verdict is DataType.DOUBLE
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/test_sqlite_probe.py]: Addressed: switched `test_text_coerce_widens_to_double` (and now the other affected tests too) to `CREATE TABLE t (v)` (BLOB affinity) so the inserted text values stay TEXT and actually exercise the coerce branch.

File: tests/test_storage_type_refinement.py [hunk 8]
```diff
@@ -391,6 +416,375 @@ def _unreachable(**kw):  # used as a wraps target only — the spy.assert_not_ca
...
+                    "version": 4,
+                    "name": "items",
+                    "sql_table": "items",
+                    "data_source": "live",
+                    "columns": [
+                        {"name": "qty", "sql": "qty", "type": "INT"},
+                    ],
+                },
+                f,
+            )
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/test_storage_type_refinement.py]: Addressed: switched the v4 fixture to `mig.CURRENT_VERSIONS["SlayerModel"] - 1` so the YAML genuinely carries a recent legacy version that preserves the modern `"INT"` token, and applied `NOSONAR(S7493)` on the sync-`open()` calls (matching the existing pattern in this file). Note: the test's INT-only branch also flipped — under the corrected `has_refineable_columns` / `has_sqlite_widenable_columns` split (Codex feedback), a missing DS for an INT-only legacy dict no longer hard-fails; it logs a WARNING and skips refinement. The test asserts the new behavior.

File: tests/test_storage_type_refinement.py [hunk 9]
```diff
@@ -497,6 +957,84 @@ async def test_missing_datasource_silent_for_text_only_model(self, tmp_path) ->
...
+        storage = YAMLStorage(base_dir=base)
+        result = _refine_one_model_for_cli(
+            inner=storage, ds_name="live", model_name="items", dry_run=True,
+        )
+        assert result is True
+        out = capsys.readouterr().out
+        assert "qty" in out
+        assert "INT" in out
+        assert "DOUBLE" in out
+
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/test_storage_type_refinement.py]: Addressed: converted `test_sqlite_int_only_legacy_dict_widens_via_cli` from `async def` to `def` (no awaits). Same for `test_missing_datasource_skips_silently_for_sqlite_int_only`.

File: tests/test_storage_type_refinement.py [hunk 10]
```diff
@@ -497,6 +957,84 @@ async def test_missing_datasource_silent_for_text_only_model(self, tmp_path) ->
...
+        storage = YAMLStorage(base_dir=base)
+        result = _refine_one_model_for_cli(
+            inner=storage, ds_name="live", model_name="items", dry_run=True,
+        )
+        assert result is False
+        # Skip message printed on stderr.
+        err = capsys.readouterr().err
+        assert "datasource 'live' unavailable" in err
+        assert "items" in err
+
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/test_storage_type_refinement.py]: Addressed (paired with thread above): converted to plain `def`.
